### PR TITLE
feat: PoSt workers

### PR DESCRIPTION
**Pull request by magik6k**
## Related Issues

Based on https://github.com/filecoin-project/lotus/pull/7600

## Proposed Changes
PoSt workers:
* Winning and Window PoSt
* Multiple PoSt workers can be connected to a lotus-miner process
  * In multi-partition PoSt, it should get run on multiple workers in parallel, up to the number of partitions
* PoSt workers can read challenges from remote storage
  * This meant that lotus-miner (nor PoSt workers) no longer need filesystem access to sectors;
* Handling of corrupted sectors should be much more robust
  * If we fail to read challenges for a sector, we'll skip it and retry without trying to create a bad post snark

## Additional Info
This PR introduces a number of new features:
* Window and Winning PoSt Workers
  * PoSt workers are specialized instances of `lotus-worker` - one lotus-worker instance can only be one of:
    * WindowPoSt worker
    * WinningPoSt worker
    * Worker for other sealing tasks
  * When a [winning/window]PoSt worker connects to lotus-miner, the miner will delegate all [winning/window]PoSt tasks to those workers - no PoSt tasks will be executed locally on the miner instance
* WindowPoSt partitions are computed in parallel
  * This should make it much easier to respond to WindowPoSt on time with multiple partitions per deadline
* Responding to PoSt challenges from remote storage
  * PoSt workers don't need local access to sectors - workers computing PoSt will ask other workers to read challenges from their storage if PoSt workers don't have access to sector storage.

### Usage
* (update lotus-miner and all workers (at the very least all workers with access to storage paths; post workers can read challanges directly from other workers, and if old workers are called, PoSt will fail))
* Have a lotus-miner running
* Start `lotus-worker run [--winningpost/--windowpost]` (use the same envvars like for any other worker)
* See the worker registered in `lotus-miner sealing workers`
* In miner info, see the PoSt worker, e.g. `Workers: Seal(1) WdPoSt(0) WinPoSt(1)`

## Followups

* Redundant PoSt - compute the same PoSt on multiple workers, submit first valid result
* Ability to entirely disable PoSt on the main lotus-miner instance - right now this happens after PoSt workers get connected

## Checklist

Before you mark the PR ready for review, please make sure that:
- [ ] All commits have a clear commit message.
- [x] The PR title is in the form of of `<PR type>: <area>: <change being made>`
    - example: ` fix: mempool: Introduce a cache for valid signatures`
    - `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_, _misc_,_perf_, _refactor_, _revert_, _style_, _test_
    - `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_
- [x] This PR has tests for new functionality or change in behaviour
- [ ] If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
- [x] CI is green

### CODE REVIEW

File: extern/sector-storage/worker_local.go [hunk 1]
```diff
@@ -508,6 +509,92 @@ func (l *LocalWorker) UnsealPiece(ctx context.Context, sector storage.SectorRef,
...
+				log.Errorf("get sector: %d, vanilla: %s, offset: %d", s.SectorNumber, vanilla)
+				return
+			}
+			vproofs[i] = vanilla
+		}(i, s)
+	}
+	wg.Wait()
+
+	if len(skipped) > 0 {
+		panic("todo") // big TODO
```
Comment (arajasek) [inline extern/sector-storage/worker_local.go:587]: Oops

**Review by arajasek — COMMENTED** [review]
Some notes

File: extern/sector-storage/manager_post.go [hunk 2]
```diff
@@ -0,0 +1,223 @@
...
+			defer wg.Done()
+
+			sectors := make([]storiface.PostSectorChallenge, 0)
+			for i := uint64(0); i < maxPartitionSize; i++ {
+				si := i + partIdx*maxPartitionSize
+				if si >= uint64(len(postChallenges.Sectors)) {
+					break
+				}
+
+				snum := postChallenges.Sectors[si]
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/manager_post.go:152]: Yep

File: extern/sector-storage/sched_post.go [hunk 3]
```diff
@@ -0,0 +1,232 @@
...
+	ps := &poStScheduler{
+		workers:  map[storiface.WorkerID]*workerHandle{},
+		postType: t,
+	}
+	ps.cond = sync.NewCond(&ps.lk)
+	return ps
+}
+
+func (ps *poStScheduler) MaybeAddWorker(wid storiface.WorkerID, tasks map[sealtasks.TaskType]struct{}, w *workerHandle) bool {
+	if _, ok := tasks[ps.postType]; !ok {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/sched_post.go:36]: CLI already has checks for this, and even if the API is called incorrectly, the worker will just ignore other task types

File: extern/sector-storage/sched_post.go [hunk 4]
```diff
@@ -0,0 +1,232 @@
...
+	ps.lk.Unlock()
+	select {
+	case <-w.closedMgr:
+	case <-time.After(time.Second):
+		log.Errorf("timeout closing worker manager goroutine %s", wid)
+	}
+	ps.lk.Lock()
+}
+
+func (ps *poStScheduler) schedClose() {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/sched_post.go:216]: It's called in `manager.go:Manager.Close`

**Comment by RobQuistNL** [comment]
Don't mind me, just leaving message here so I can stalk the progress :)

**Comment by magik6k** [comment]
(marking as draft until the skipping issue is figured out)

**Comment by magik6k** [comment]
Just tested in lotus-pond, and it works really well:
* Both PoSt types work
* Disputer doesn't dispute my PoSts
* After restarting lotus-miner, the worker auto-reconnects
* When the worker is stopped, lotus-miner switches back to local PoSt automatically

File: extern/sector-storage/worker_local.go [hunk 5]
```diff
@@ -559,6 +560,99 @@ func (l *LocalWorker) UnsealPiece(ctx context.Context, sector storage.SectorRef,
...
+	wg.Add(len(sectors))
+
+	vproofs := make([][]byte, len(sectors))
+	var rerr error
+
+	for i, s := range sectors {
+		go func(i int, s storiface.PostSectorChallenge) {
+			defer wg.Done()
+
+			// todo context with tighter deadline (+config)
```
Comment (ZenGround0) [inline extern/sector-storage/worker_local.go:580]: Does this todo need to be done before merging?

File: extern/sector-storage/worker_local.go [hunk 6]
```diff
@@ -559,6 +560,99 @@ func (l *LocalWorker) UnsealPiece(ctx context.Context, sector storage.SectorRef,
 	})
 }
 
+func (l *LocalWorker) GenerateWinningPoSt(ctx context.Context, ppt abi.RegisteredPoStProof, mid abi.ActorID, sectors []storiface.PostSectorChallenge, randomness abi.PoStRandomness) ([]proof.PoStProof, error) {
+	sb, err := l.executor()
+	if err != nil {
+		return nil, err
+	}
+
+	// todo throttle config
```
Comment (ZenGround0) [inline extern/sector-storage/worker_local.go:569]: Does this todo need to be done before merging?

File: extern/sector-storage/manager_post.go [hunk 7]
```diff
@@ -0,0 +1,241 @@
...
+		return nil
+	})
+
+	log.Warnf("generateWindowPost partition:%d, get skip count:%d", partIndex, len(result.Skipped))
+
+	return result.PoStProofs, result.Skipped, err
+}
+
+func (m *Manager) GenerateWinningPoStWithVanilla(ctx context.Context, proofType abi.RegisteredPoStProof, minerID abi.ActorID, randomness abi.PoStRandomness, proofs [][]byte) ([]proof.PoStProof, error) {
+	//TODO implement me
```
[the remark this answers is not in the document]
Comment (ZenGround0) [reply extern/sector-storage/manager_post.go:234]: I think I get it now, this needs the same interface as the ffiwrapper which needs these.

**Comment by ZenGround0** [comment]
@magik6k since we're using a different pathway through ffi to do post in the case a worker is hooked up it seems like lotus-bench should adapt to be able to exercise the new path.  I'm not sure whether the typical lotus-bench use case is 1) exercise the operator's specific setup or 2) exercise general purpose proving pathways on the miner.

If its 1) we could hook up lotus-bench to the manager so it can talk to workers.  If its 2) I think we should add a lotus-bench check that does the challenge -> vanilla proof -> snark ( -> merge multi partition?)

This may be overkill because ffi is doing similar things inside for each.  What do you think?

File: extern/sector-storage/manager_post.go [hunk 8]
```diff
@@ -0,0 +1,241 @@
...
+	}
+
+	spt := sectorInfo[0].SealProof
+
+	ppt, err := spt.RegisteredWinningPoStProof()
+	if err != nil {
+		return nil, err
+	}
+
+	postChallenges, err := ffi.GeneratePoStFallbackSectorChallenges(ppt, minerID, randomness, sectorNums)
```
[the remark this answers is not in the document]
Comment (ZenGround0) [reply extern/sector-storage/manager_post.go:47]: I think I understand this now, the ffi exposes several different layers of abstraction

File: extern/sector-storage/faults.go [hunk 9]
```diff
@@ -161,25 +99,4 @@ func (m *Manager) CheckProvable(ctx context.Context, pp abi.RegisteredPoStProof,
 	return bad, nil
 }
 
-func addCachePathsForSectorSize(chk map[string]int64, cacheDir string, ssize abi.SectorSize) {
-	switch ssize {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/faults.go:165]: Yeah, if we can read challanges, even random ones, that probably means that the sector is fine (when challenges are read, they are also validated, at least that's what I've been told)

File: extern/sector-storage/storiface/resources.go [hunk 10]
```diff
@@ -463,6 +463,104 @@ var ResourceTable = map[sealtasks.TaskType]map[abi.RegisteredSealProof]Resources
 			MaxParallelism: 1,
 			GPUUtilization: 1.0,
 
+			BaseMinMemory: 8 << 20,
+		},
+	},
+	sealtasks.TTGenerateWindowPoSt: {
+		abi.RegisteredSealProof_StackedDrg64GiBV1: Resources{
+			MaxMemory: 120 << 30, // TODO: Confirm
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/storiface/resources.go:471]: Yep

File: extern/sector-storage/manager_post.go [hunk 11]
```diff
@@ -0,0 +1,241 @@
...
+	}
+
+	spt := sectorInfo[0].SealProof
+
+	ppt, err := spt.RegisteredWindowPoStProof()
+	if err != nil {
+		return nil, nil, err
+	}
+
+	maxPartitionSize, err := builtin.PoStProofWindowPoStPartitionSectors(ppt) // todo proxy through chain/actors
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/manager_post.go:118]: Probably fine for now, if we were to touch anything PoSt related, a lot of this code will see significant changes / review anyways.

File: cmd/lotus-miner/info.go [hunk 12]
```diff
@@ -343,16 +344,45 @@ func handleMiningInfo(ctx context.Context, cctx *cli.Context, fullapi v0api.Full
...
+
+		var nseal, nwdpost, nwinpost int
+
+	wloop:
+		for _, st := range ws {
+			if !st.Enabled {
+				continue
+			}
+
+			for _, task := range st.Tasks {
```
[the remark this answers is not in the document]
Comment (magik6k) [reply cmd/lotus-miner/info.go:367]: Should be addressed now

File: extern/sector-storage/worker_local.go [hunk 13]
```diff
@@ -607,17 +625,34 @@ func (l *LocalWorker) GenerateWindowPoSt(ctx context.Context, ppt abi.Registered
...
 	var skipped []abi.SectorID
 
-	// todo throttle config
 	var wg sync.WaitGroup
 	wg.Add(len(sectors))
 
 	vproofs := make([][]byte, len(sectors))
 
 	for i, s := range sectors {
+		if l.challengeThrottle != nil {
```
Comment (ZenGround0) [inline extern/sector-storage/worker_local.go:634]: I think I'm missing something stupid but won't this block on the first sector?  The go routines below won't populate the channel for unblocking until they are spawned but the main loop needs to make it past this first to kick them off right?

**Review by ZenGround0 — APPROVED** [review]
The question about lotus bench remains and I didn't review the new worker itest code too carefully but I'm happy with this going in as is.

File: extern/sector-storage/worker_local.go (hunk 13 above)
Comment (magik6k) [reply extern/sector-storage/worker_local.go:634]: Oh, the channel ops are reversed

File: extern/sector-storage/sched_post.go [hunk 14]
```diff
@@ -0,0 +1,232 @@
...
+				continue
+			case <-worker.closingMgr:
+				return
+			}
+		}
+
+		if storiface.WorkerID(curSes) != wid {
+			if curSes != ClosedWorkerID {
+				// worker restarted
+				log.Warnw("worker session changed (worker restarted?)", "initial", wid, "current", curSes)
```
[the remark this answers is not in the document]
Comment (magik6k) [reply extern/sector-storage/sched_post.go:191]: Yes, that delete is enough

**Comment by magik6k** [comment]
On lotus-bench + setting good resource numbers - I think this can be done after this lands in master; I'd love to stop making this PR bigger, especially given that it seems to work pretty well already

(opened https://github.com/filecoin-project/lotus/pull/8373 with some in-progress work on top of this PR)

**Review by arajasek — APPROVED** [review]
👍 on the last five commits

File: extern/sector-storage/faults.go (hunk 9 above)
[the remark this answers is not in the document]
Comment (yhio) [reply extern/sector-storage/faults.go:165]: Will it seriously increase the time spent on check?

**Comment by lbj2004032** [comment]
x

> ## Related Issues
> Based on #7600
> 
> ## Proposed Changes
> PoSt workers:
> 
> * Winning and Window PoSt
> * Multiple PoSt workers can be connected to a lotus-miner process
>   
>   * In multi-partition PoSt, it should get run on multiple workers in parallel, up to the number of partitions
> * PoSt workers can read challenges from remote storage
>   
>   * This meant that lotus-miner (nor PoSt workers) no longer need filesystem access to sectors;
> * Handling of corrupted sectors should be much more robust
>   
>   * If we fail to read challenges for a sector, we'll skip it and retry without trying to create a bad post snark
> 
> ## Additional Info
> This PR introduces a number of new features:
> 
> * Window and Winning PoSt Workers
>   
>   * PoSt workers are specialized instances of `lotus-worker` - one lotus-worker instance can only be one of:
>     
>     * WindowPoSt worker
>     * WinningPoSt worker
>     * Worker for other sealing tasks
>   * When a [winning/window]PoSt worker connects to lotus-miner, the miner will delegate all [winning/window]PoSt tasks to those workers - no PoSt tasks will be executed locally on the miner instance
> * WindowPoSt partitions are computed in parallel
>   
>   * This should make it much easier to respond to WindowPoSt on time with multiple partitions per deadline
> * Responding to PoSt challenges from remote storage
>   
>   * PoSt workers don't need local access to sectors - workers computing PoSt will ask other workers to read challenges from their storage if PoSt workers don't have access to sector storage.
> 
> ### Usage
> * (update lotus-miner and all workers (at the very least all workers with access to storage paths; post workers can read challanges directly from other workers, and if old workers are called, PoSt will fail))
> * Have a lotus-miner running
> * Start `lotus-worker run [--winningpost/--windowpost]` (use the same envvars like for any other worker)
> * See the worker registered in `lotus-miner sealing workers`
> * In miner info, see the PoSt worker, e.g. `Workers: Seal(1) WdPoSt(0) WinPoSt(1)`
> 
> ## Followups
> * Redundant PoSt - compute the same PoSt on multiple workers, submit first valid result
> * Ability to entirely disable PoSt on the main lotus-miner instance - right now this happens after PoSt workers get connected
> 
> ## Checklist
> Before you mark the PR ready for review, please make sure that:
> 
> * [ ]  All commits have a clear commit message.
> * [x]  The PR title is in the form of of `<PR type>: <area>: <change being made>`
>   
>   * example: ` fix: mempool: Introduce a cache for valid signatures`
>   * `PR type`: _fix_, _feat_, _INTERFACE BREAKING CHANGE_, _CONSENSUS BREAKING_, _build_, _chore_, _ci_, _docs_, _misc_,_perf_, _refactor_, _revert_, _style_, _test_
>   * `area`: _api_, _chain_, _state_, _vm_, _data transfer_, _market_, _mempool_, _message_, _block production_, _multisig_, _networking_, _paychan_, _proving_, _sealing_, _wallet_
> * [x]  This PR has tests for new functionality or change in behaviour
> * [ ]  If new user-facing features are introduced, clear usage guidelines and / or documentation updates should be included in https://lotus.filecoin.io or [Discussion Tutorials.](https://github.com/filecoin-project/lotus/discussions/categories/tutorials)
> * [x]  CI is green

seperate

**Comment by lbj2004032** [comment]
What would happen when miner process exit?
I thought minners prefer a framework that wndpost as a standalone process,
does wndpost/winpost run seperate from minner?

**Comment by lbj2004032** [comment]
> What would happen when miner process exit?
> I thought minners prefer a framework that wndpost as a standalone process,
> does wndpost/winpost run seperate from minner?

**Comment by BBQFIL** [comment]
@magik6k 
Hello, I want to ask how miner judges that a post worker has been closed.
I encountered some problems today. When one of my post workers lost connection to the miner due to network card failure, the miner still handed over the partition to the lost post worker for processing. This is not in line with my expectations. In my opinion, when a post worker fails to handle work, miner should assign tasks to other healthy post workers.

image

**Pull request merged**